### PR TITLE
refactor: Remove unused includes from wallet.cpp

### PR DESCRIPTION
**Pull request by maflcko**
This makes compilation of wallet.cpp use a few % less memory and time, locally.

Created in the context of https://github.com/bitcoin/bitcoin/issues/28109, but I don't think it is enough to actually fix this problem.

### CODE REVIEW

**Comment by theuni** [comment]
> Good job! The circular dependency "wallet/fees -> wallet/wallet -> wallet/fees" is no longer present.
> Please remove it from EXPECTED_CIRCULAR_DEPENDENCIES in test/lint/lint-circular-dependencies.py
> to make sure this circular dependency is not accidentally reintroduced.

🎉

**Review by ryanofsky — APPROVED** [review]
Code review ACK ccccbd8dd0f5cd74a744e74608c4ea102ec33fd1

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -5,50 +5,72 @@
 
 #include <wallet/wallet.h>
 
+#include <bitcoin-config.h>
```
Comment (hebasto) [inline src/wallet/wallet.cpp:8]:
```suggestion
#if defined(HAVE_CONFIG_H)
#include <config/bitcoin-config.h>
#endif
```

**Review by hebasto — APPROVED** [review]
ACK fa8940852f88f74cb0d4975e05ec32fd3594961c

File: src/txmempool.h [hunk 2]
```diff
@@ -6,27 +6,17 @@
...
-
 #include <coins.h>
 #include <consensus/amount.h>
 #include <indirectmap.h>
 #include <kernel/cs_main.h>
-#include <kernel/mempool_entry.h>
+#include <kernel/mempool_entry.h>          // IWYU pragma: export
+#include <kernel/mempool_limits.h>         // IWYU pragma: export
+#include <kernel/mempool_options.h>        // IWYU pragma: export
+#include <kernel/mempool_removal_reason.h> // IWYU pragma: export
```
Comment (hebasto) [inline src/txmempool.h:16]: It seems `IWYU pragma: export` causes:
```
txmempool.h should add these lines:
...
#include "kernel/mempool_limits.h"
#include "kernel/mempool_options.h"
#include "kernel/mempool_removal_reason.h"                   // for MemPoolRemovalReason (ptr only)
...
```

Perhaps, a better solution is to use `contrib/devtools/iwyu/bitcoin.core.imp`.

**Comment by fanquake** [comment]
> Created in the context of https://github.com/bitcoin/bitcoin/issues/28109, but I don't think it is enough to actually fix this problem.

It's a step in the right direction, we're now failing on `45934` sections, compared to `46006` in #28109. However we're still a long way north of the limit, which as I understand it, is `32768`.

File: src/txmempool.h (hunk 2 above)
Comment (hebasto) [reply src/txmempool.h:16]: > I think this bug should be reported to IWYU...

https://github.com/include-what-you-use/include-what-you-use/issues/1280

**Comment by theuni** [comment]
Concept ACK. I didn't review the includes changes in detail, but the move makes sense to me. And I'm always up for one-data-structure-per-header :)

**Comment by Empact** [comment]
Concept ACK

nit: IMO better to split the changes from the iwyu adds/removals in the spirit of the "[Pull Request Philosophy](https://github.com/bitcoin/bitcoin/blob/b7138252ace6d21476964774e094ed1143cd7a1c/CONTRIBUTING.md#pull-request-philosophy)"

**Comment by Empact** [comment]
To be clear as to what I'm referring to, note your commit messages both end with "Also ...". This alludes to the fact that you're doing a variety of things in each, which is a burden (in this case a minor one) on review.

Feel free to disregard as a nit, but I think this is a good practice to keep in mind, that we may maintain discipline in how we structure our changes for the benefit of review.

**Comment by RandyMcMillan** [comment]
Concept ACK

Screen Shot 2023-08-07 at 1 49 13 PM


macOS Catalina Version 10.15.7 (19H2026)

**Comment by TheCharlatan** [comment]
Re https://github.com/bitcoin/bitcoin/pull/28200#issuecomment-1668087768

> Can you post a patch that compiles?

Seems like an easy win, no?

```diff
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
index 7c74b6b9ff..fe4c3d0f67 100644
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -79,6 +79,8 @@
 #include <tuple>
 #include <variant>
 
+#include <boost/signals2/signal.hpp>
+
 struct KeyOriginInfo;
 
 using interfaces::FoundBlock;
diff --git a/src/wallet/wallet.h b/src/wallet/wallet.h
index 8a1a513062..24cedd5160 100644
--- a/src/wallet/wallet.h
+++ b/src/wallet/wallet.h
@@ -49,8 +49,6 @@
 #include <utility>
 #include <vector>
 
-#include <boost/signals2/signal.hpp>
-
 class CKey;
 class CKeyID;
 class CPubKey;
```

**Review by sedited — APPROVED** [review]
ACK fac3e2fa6adf9cb034ea5ee28c572b25cc7e3c2d

Re https://github.com/bitcoin/bitcoin/pull/28200#issuecomment-1669483448

> No. The boost include is still there (via src/wallet/scriptpubkeyman.h:#include <boost/signals2/signal.hpp>

Yeah, totally. Should have checked this before trusting the iwyu suggestion.

**Review by hebasto — APPROVED** [review]
ACK fa6286891fa4164510e4fbf4bc214ce3033b2d1b, I have reviewed the code and it looks OK.

**Pull request merged**